nemo/lightning/run/plugins.py

```diff
@@ -241,3 +242,65 @@ def setup(self, task: run.Partial | run.Script, executor: run.Executor):
             if isinstance(executor, run.SlurmExecutor):
                 assert task.trainer.num_nodes == executor.nodes
                 assert task.trainer.devices == executor.nproc_per_node()
+
+
+@dataclass(kw_only=True)
+class PerfEnvPlugin(run.Plugin):
+    """
+    A plugin for setting up performance optimized environments.
+
+    Attributes:
+        enable_layernorm_sm_margin (bool): Set SM margin for TransformerEngine's Layernorm, so
+            in order to not block DP level communication overlap.
+        enable_vboost (bool): Whether to steer more power towards tensor cores via
+            `sudo nvidia-smi boost-slider --vboost 1`. May not work on all systems.
+    """
+
+    enable_layernorm_sm_margin: bool = True
+    enable_vboost: bool = False
+
+    def get_vboost_srun_cmd(self, nodes, job_dir):
+        import shlex
+
+        vboost_cmd = " ".join(
+            [
+                "\n# Command 0: enable vboost\n\n",
+                "srun",
+                f"--ntasks={nodes}",
+                "--output",
+                os.path.join(job_dir, "vboost.out"),
+                "--error",
+                os.path.join(job_dir, "vboost.err"),
+                "bash -c ",
+                shlex.quote("sudo nvidia-smi boost-slider --vboost 1"),
+            ],
+        )
+
+        return vboost_cmd
+
+    def setup(self, task: run.Partial | run.Script, executor: run.Executor):
+
+        if task.trainer.strategy.__fn_or_cls__ == MegatronStrategy:
+            # Force program order kernel launch for TP, CP overlap
+            tp_size = task.trainer.strategy.tensor_model_parallel_size
+            cp_size = task.trainer.strategy.context_parallel_size
+            if tp_size > 1 and cp_size > 1:
+                executor.env_vars["CUDA_DEVICE_MAX_CONNECTIONS"] = 1
+
+            # Set LayerNorm SM margin to support the overlap with LayerNorm kernel
+            if self.enable_layernorm_sm_margin:
+                executor.env_vars["NVTE_FWD_LAYERNORM_SM_MARGIN"] = 8
+                executor.env_vars["NVTE_BWD_LAYERNORM_SM_MARGIN"] = 8
```

Comment:
Can we just make this to 16?
This should cover all NIC cases used for data-parallel collectives.

Reply:
Okay, I also added a knob to specify the margin